src/pages/workspaces/hooks/useDeletionPolling.ts

```diff
@@ -0,0 +1,137 @@
+import _ from 'lodash/fp';
+import { useEffect, useState } from 'react';
+import { Ajax } from 'src/libs/ajax';
+import { workspacesStore, workspaceStore } from 'src/libs/state';
+import { pollWithCancellation } from 'src/libs/utils';
+import { WorkspaceState, WorkspaceWrapper as Workspace } from 'src/libs/workspace-utils';
+
+const updateWorkspacesList = (
+  workspaces: Workspace[],
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): Workspace[] =>
+  workspaces.map((ws) => {
+    if (ws.workspace.workspaceId === workspaceId) {
+      const update = _.cloneDeep(ws);
+      update.workspace.state = state;
+      update.workspace.errorMessage = errorMessage;
+      return update;
+    }
+    return ws;
+  });
+
+const updateWorkspace = (
+  workspace: Workspace,
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): Workspace => {
+  if (workspace.workspace.workspaceId === workspaceId) {
+    const update = _.cloneDeep(workspace);
+    update.workspace.state = state;
+    update.workspace.errorMessage = errorMessage;
+    return update;
+  }
+  return workspace;
+};
+
+export const useDeletetionPolling = (workspaces: Workspace[]) => {
+  // we have to do the signal/abort manually instead of with useCancelable so that the it can be cleaned up in the
+  // this component's useEffect, instead of the useEffect in useCancelable
+  // const workspaces: Workspace[] = useStore<Workspace[]>(workspacesStore)
+  const [controller, setController] = useState(new window.AbortController());
+  const abort = () => {
+    controller.abort();
+    setController(new window.AbortController());
+  };
+
+  useEffect(() => {
+    const checkWorkspaceDeletion = async (workspace: Workspace) => {
+      const doUpdate = (state: WorkspaceState, errorMessage?: string) => {
+        const workspaceId = workspace.workspace.workspaceId;
+        abort();
+        workspacesStore.update((wsList) => updateWorkspacesList(wsList, workspaceId, state, errorMessage));
+        workspaceStore.update((ws) => updateWorkspace(ws, workspaceId, state, errorMessage));
+      };
+
+      try {
+        const wsResp: Workspace = await Ajax(controller.signal)
+          .Workspaces.workspace(workspace.workspace.namespace, workspace.workspace.name)
+          .details(['workspace.state', 'workspace.errorMessage']);
+        const state = wsResp.workspace.state;
+
+        if (state === 'DeleteFailed') {
```

Comment:
Why just `DeleteFailed` instead of anything but `Deleting`?

Reply:
Mostly just because I was mentally framing it in terms of what the next possible states could be: the only thing it ever can transition to from 'Deleting' is 'DeleteFailed' or a 404.

Making it more broad is probably better though.